src/main/java/io/r2dbc/postgresql/client/SSLSessionHandlerAdapter.java

```diff
@@ -65,13 +74,14 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) {
         }
     }
 
-    private void processSslDisabled() {
+    private void processSslDisabled(ChannelHandlerContext ctx) {
         if (this.sslConfig.getSslMode().requireSsl()) {
             PostgresqlSslException e =
                 new PostgresqlSslException("Server support for SSL connection is disabled, but client was configured with SSL mode " + this.sslConfig.getSslMode());
             completeHandshakeExceptionally(e);
         } else {
             completeHandshake();
+            ctx.channel().pipeline().remove(this);
```

Comment:
Noticed this path as well, which could be reached if the server is `ssl=off` and the client is `sslmode=prefer`.  Probably not tested currently, but the adapter would still be in the pipeline and fail subsequent reads.